HIP/hip-0000-atomic-association-contracts.md

```diff
@@ -0,0 +1,104 @@
+---
+hip: 0000
+title: Waived association requirement for atomic transfers for smart contracts
+author: Matthew DeLorenzo <mdelore.ufl@gmail.com>, Vae Vecturne <vea.vecturne@gmail.com>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+created: 2023-04-06
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/696
+---
+
+## Abstract
+
+Smart contracts are required to be associated with all tokens they custody. This makes sense when the contract has a long lived balance > 0; Hedera state is expanded and must be paid for in ongoing maintenance and storage costs to the network. However, there are many use cases that call for smart contracts atomically taking custody of a token, often to transfer to another entity within the same contract call. In this case, association is still required but Hedera state is not expanded.
+
+We propose that association not be required for Hedera entities that only custody a token during an atomic transaction.
+
+## Motivation
+
+Currently, a smart contract attempting to custody an unassociated token will be immediately rejected by the network. This presents a problem in some decentralized finance (DeFi) applications that use smart contracts to atomically custody tokens. Decentraliztion may be achieved by allowing contracts to remain token agnostic, but also prevent spamming attacks by allowing any user to associate tokens to a DeFi smart contract. 
```

Comment:
Does the proposal refer to all smart contract regardless of creation path?
For example those created via HAPI ContractCreate have the option of setting [max_automatic_token_associations](https://github.com/hashgraph/hedera-protobufs/blob/main/services/contract_create.proto#L173) and updating it in a [ContractUpdate](https://github.com/hashgraph/hedera-protobufs/blob/main/services/contract_update.proto#L105).

This might be a good section to address why that path is not sufficient for the proposal

Reply:
This proposal refers to all smart contracts regardless of creation path. Smart contracts are often intermediaries for many different tokens, so it becomes a challenge to maintain association maps and retain decentralization. Presently, dexes are associating/dissociating tokens using precompiles every time they are used as intermediaries. If Hedera was able to refine association checks such that the ending contract balance must be greater than zero, it would eliminate the need to pay 5 cents more per transaction.

We will address why the max_automatic_token_association is not sufficient.

Reply:
@Nana-EC this is now addressed on line 24